File: vb.net/RegistrationForm/Validation/FormValidator.vb
Imports System.Text.RegularExpressions

Public Class FormValidator
    Private Shared ReadOnly MAX_LENGTH As Integer = 254
    Private Shared ReadOnly NOT_SPECIFIED = "N/A"


    Public Shared Function UsernameValidator(input As TextBox) As String
        Dim name = input.Text()
        Dim pattern As String = "[A-Za-z]+"
        Dim regex As New Regex(pattern)

        If String.IsNullOrWhiteSpace(name) Then
            Throw New ArgumentException("Name cannot be empty")
        End If

        If Not regex.IsMatch(name) Then
            Throw New ArgumentException("Invalid Name")
        End If

        If name.Length() > MAX_LENGTH Then
            Throw New ArgumentException($"Name exceeds maximum length of {MAX_LENGTH} characters")
        End If

        Return name.Trim()

    End Function

    Public Shared Function EmailValidation(input As TextBox) As String
        Dim email = input.Text()
        Dim pattern As String = "^[^@\s]+@[^@\s]+\.[^@\s]+$"
        Dim regex As New Regex(pattern)


        If String.IsNullOrWhiteSpace(email) Then
            Throw New ArgumentException("Email address cannot be empty")
        End If

        If Not regex.IsMatch(email) Then
            Throw New ArgumentException("Invalid email address format")
        End If

        If email.Length > MAX_LENGTH Then
            Throw New ArgumentException($"Email address exceeds maximum length of {MAX_LENGTH} characters")
        End If

        Return email.Trim()

    End Function

    Public Shared Function PasswordValidation(inputA As TextBox, inputB As TextBox) As String
        Dim pwd As String = inputA.Text()
        Dim cPwd As String = inputB.Text()
        Dim MIN_LENGTH As Integer = 5


        If String.IsNullOrWhiteSpace(pwd) Or String.IsNullOrWhiteSpace(cPwd) Then
            Throw New ArgumentException("Fields password or confirm password is empty")
        End If

        If pwd.Length < MIN_LENGTH Then
            Throw New ArgumentException("Please enter a longer password.")
        End If

        If Not pwd.Equals(cPwd) Then
            Throw New ArgumentException("Fields password or confirm password do not match. Please ensure that you enter the same password in both fields")
        End If

        Return pwd

    End Function

    ''' <summary>
    ''' Return an integer
    ''' </summary>
    ''' <param name="rb_a">Male</param>
    ''' <param name="rb_b">Female</param>
    ''' <param name="rb_c">Other</param>
    ''' <returns></returns>
    Public Shared Function GetSelectedGender(rb_a As RadioButton, rb_b As RadioButton, rb_c As RadioButton) As Integer

        If rb_a.Checked Then
            Return GenderType.Male
        End If

        If rb_b.Checked Then
            Return GenderType.Female
        End If
        If rb_c.Checked Then
            Return GenderType.Other
        End If

        Return GenderType.Unsepecified


    End Function


    Public Shared Function AddressValidator(input As TextBox) As String

        If String.IsNullOrWhiteSpace(input.Text) Then
            Return NOT_SPECIFIED
        End If

        If input.TextLength() > MAX_LENGTH Then
            Throw New ArgumentException($"Address exceeds maximum length of {MAX_LENGTH} characters")
        End If

        Return input.Text().Trim()

    End Function

    Public Shared Function GetSelectedState(input As ComboBox) As Integer

        Dim selectedIndex = input.SelectedIndex

        If selectedIndex = StatesType.SELECT_STATE Then
            Return StatesType.UNSPECIFIED
        End If

        Return selectedIndex

    End Function

    Public Shared Function CountryValidation(input As TextBox) As String
        Dim country As String = input.Text()
        Dim pattern As String = "[A-Za-z]+"
        Dim regex As New Regex(pattern)

        If String.IsNullOrWhiteSpace(country) Then
            Return NOT_SPECIFIED
        End If

        If Not regex.IsMatch(country) Then
            Throw New ArgumentException("Invalid Country Format")
        End If

        If country.Length() > MAX_LENGTH Then
            Throw New ArgumentException($"Country exceeds maximum length of {MAX_LENGTH} characters")
        End If

        Return country.Trim()
    End Function

    Public Shared Function PhoneNumberValidation(input As MaskedTextBox) As String

        If String.IsNullOrWhiteSpace(input.Text) Then
            Return NOT_SPECIFIED
        End If
        Const MAX_CHAR = 12
        If Not input.TextLength = MAX_CHAR Then
            Throw New ArgumentException("Enter your full phone number")
        End If

        Return input.Text.Trim()
    End Function

End Class

File: java/RegistrationForm/src/validators/FormValidator.java
package validators;

import model.Gender;
import model.States;

import javax.swing.*;
import java.util.Arrays;
import java.util.regex.Matcher;
import java.util.regex.Pattern;


/**
 * Validates the fields from the form
 */
public class FormValidator {
    private static final String NOT_SPECIFIED = "N/A";
    private static final int MAX_LENGTH = 120;


    /**
     * @param input name field
     * @return name
     * @throws IllegalArgumentException if the input is not well entered
     */
    public static String nameValidation(JTextField input) {
        Pattern stringVerification = Pattern.compile("[A-Za-z]+");
        String name = input.getText();
        Matcher matcher = stringVerification.matcher(name);
        if(!matcher.matches()){
            throw new IllegalArgumentException("Field 'name' was not entered correctly");
        }
        if(name.length() > MAX_LENGTH){
            throw new IllegalArgumentException("Field 'name' exceeds the maximum length by " + (name.length() - MAX_LENGTH));
        }
        return name;
    }

    /**
     * @param input email field
     * @return email
     * @throws IllegalArgumentException if the input is not well entered
     */
    public static String emailValidation(JTextField input) {
        String email = input.getText();
        Pattern stringVerification = Pattern.compile("(?!\\.)[\\w\\-_.]*[^.](@\\w+)(\\.\\w+(\\.\\w+)?[^.\\W])");
        Matcher matcher = stringVerification.matcher(email);
        if(!matcher.matches()){
            throw new IllegalArgumentException("Field 'email' was not entered correctly");
        }
        if(email.length() > MAX_LENGTH){
            throw new IllegalArgumentException("Field 'email' exceeds the maximum length by " + (email.length() - MAX_LENGTH));
        }
        return email;
    }

    /**
     * @param inputA password field
     * @param inputB password field
     * @return password
     * @throws IllegalArgumentException if the input don't match
     */
    public static String passwordValidation(JPasswordField inputA, JPasswordField inputB) {
        char[] password = inputA.getPassword();
        char[] cPassword = inputB.getPassword();

        final int MINIMUM_LENGTH = 5;
        if(password.length < MINIMUM_LENGTH){
            throw new IllegalArgumentException("Field 'password' please enter a longer password. To ensure that your information is not stolen.");
        }
        if(!Arrays.equals(password, cPassword)){
            throw new IllegalArgumentException("Fields 'password' do not match. Please ensure that you enter the same password in both fields");
        }

        return  Arrays.toString(password).replaceAll("\\[|]|,|\\s", "");
    }

    /**
     * @param male male field
     * @param female female field
     * @return phone Number
     */
    public static Gender getGenderSelected(JRadioButton male, JRadioButton female){
        if (male.isSelected()){
            return Gender.Male;
        }
        if (female.isSelected()) {
            return Gender.Female;
        }

        return Gender.Other;

    }

    /**
     * @param box combo
     * @return the state selected
     */
    public static String getSelectedState(JComboBox<String> box){
        final int NOT_ALLOWED = 0;
        int selectedIndex = box.getSelectedIndex();
        if(selectedIndex != NOT_ALLOWED){
            return States.getName((String)box.getSelectedItem());
        }
        return null;
    }

    /**
     * @param input address field
     * @return address
     */
    public static String addressValidation(JTextField input) {
        String address = input.getText().trim();
        if(address.isEmpty()){
            return NOT_SPECIFIED;
        }

        if(address.length() > MAX_LENGTH){
            throw new IllegalArgumentException("Field 'address' exceeds the maximum length by " + (address.length() - MAX_LENGTH));
        }
        return address;
    }

    /**
     * @param input country field
     * @return country
     * @throws IllegalArgumentException if the input is not well entered
     */
    public static String countryValidation(JTextField input) {
        String country = input.getText().trim();
        Pattern stringVerification = Pattern.compile("[A-Za-z]+");
        Matcher matcher = stringVerification.matcher(country);
        if(country.isEmpty()){
            return NOT_SPECIFIED;
        }
        if(!matcher.matches()){
            throw new IllegalArgumentException("Field 'country' was not entered correctly");
        }

        if(country.length() > MAX_LENGTH){
            throw new IllegalArgumentException("Field 'country' exceeds the maximum length by " + (country.length() - MAX_LENGTH));
        }
        return country;
    }

    /**
     * @param input formatted field
     * @return phone Number
     * @throws IllegalArgumentException if the input is not well entered
     */
    public static String phoneNumberValidation(JFormattedTextField input) {
        String phoneNumber = input.getText().trim();
        if (phoneNumber.isEmpty()){
            return NOT_SPECIFIED;
        }
        if(phoneNumber.length() != 12) {
            throw new IllegalArgumentException("Field 'phone number' please enter your full phone number");
        }
        return phoneNumber;
    }

}
